idseq_pipeline/commands/accessionid2seq.py

```diff
@@ -94,7 +125,61 @@ def generate_alignment_viz_json(nt_file, nt_loc_db, db_type,
             for (species_id, species_dict) in genus_dict.iteritems():
                 with open("%s/%s.species.%d.align_viz.json" %(output_json_dir, db_type.lower(), int(species_id)), 'wb') as outjf:
                     json.dump(species_dict, outjf)
-    return "Read2Seq Size: %d, M8 lines %d" % (len(read2seq),line_count)
+    return "Read2Seq Size: %d, M8 lines %d, %d unique accession ids" % (len(read2seq),line_count, len(accession2seq))
+
+def get_sequences_by_accession_list_s3(accession_id_list, nt_loc_dict, nt_s3_path):
+    threads = []
+    error_flags = {}
+    mutex = threading.RLock()
+    semaphore = threading.Semaphore(16)
+    for accession_id, accession_info in accession_id_list.iteritems():
+	semaphore.acquire()
+	t = threading.Thread(
+	    target=get_sequence_for_thread,
+	    args=[error_flags, accession_info, accession_id, nt_loc_dict, nt_s3_path, semaphore, mutex]
+	)
+        t.start()
+        threads.append(t)
+    for t in threads:
+        t.join()
+    if error_flags:
+        raise "Sorry there was an error"
+
+def get_sequence_for_thread(error_flags, accession_info, accession_id, nt_loc_dict, nt_s3_path, semaphore, mutex):
+    try:
+        accession_info['seq'] = get_sequence_by_accession_id(accession_id, nt_loc_dict, None, nt_s3_path)
+    except:
+	mutex.acquire()
+        error_flags["error"] = 1
+        mutex.release()
+    finally:
+        semaphore.release()
```

Comment:
It's good to also protect with mutex your access to the `accession_info` dict, as it is being written to by multiple threads.   I suggest this.
```
    try:
        result = accession_info['seq'] = get_sequence_by_accession_id(accession_id, nt_loc_dict, None, nt_s3_path)
        with mutex:
            accession_info['seq'] = result
    except:
        with mutex:
            if not error_flags:
                # print the error only from one thread
                traceback.print_exc()
            error_flags["error"] = 1
            return
    finally:
        semaphore.release()
```

Reply:
Cool.  Thanks!